lib/absinthe/federation/schema/entities_field.ex

```diff
@@ -83,120 +83,141 @@ defmodule Absinthe.Federation.Schema.EntitiesField do
-       do: resolve_reference(fields[:_resolve_reference], parent, representation, resolution)
+  defp reverse_values_and_errors(res) do
+    res
+    |> Map.update!(:value, &Enum.reverse/1)
+    |> Map.update!(:errors, &Enum.reverse/1)
+  end
 
-  defp resolve_representation(_schema_type, _parent, representation, _schema),
-    do:
-      {:error,
-       "The _entities resolver tried to load an entity for type '#{Map.get(representation, "__typename")}', but no object type of that name was found in the schema"}
+  defp set_final_state(res) do
+    if Enum.all?(res.path, &(&1.state == :resolved)) do
+      Map.put(res, :state, :resolved)
+    else
+      paths = Enum.map(res.path, &%{state: &1.state, value: &1.value, errors: &1.errors})
+      raise "Expected all fields to be resolved, but got: #{paths}"
+    end
+  end
 
-  defp resolve_reference(nil, _parent, representation, %{context: context} = _resolution) do
-    args = convert_keys_to_atom(representation, context)
+  # These are the fields to be threaded through every single representation resolution.
+  defp resolution_accumulator(resolution) do
+    %{resolution | path: [], errors: [], value: []}
+  end
 
-    fn _, _ -> {:ok, args} end
+  # Resolve a single representation and accumulate it to field resolution
+  # under path key. If a field is already resolved, do not run any middlewares.
+  defp resolve_field_representation(representation, acc) do
+    if representation.state == :resolved do
+      new_path = acc.path ++ [representation]
+
+      acc
+      |> Map.put(:path, new_path)
+    else
+      local_res =
+        representation
+        |> Map.put(:context, acc.context)
+        |> Map.put(:acc, acc.acc)
+
+      result = reduce_resolution(local_res)
+      new_path = acc.path ++ [result]
+
+      acc
+      |> Map.put(:context, result.context)
+      |> Map.put(:acc, result.acc)
+      |> Map.put(:path, new_path)
+    end
   end
 
-  defp resolve_reference(
-         %{middleware: middleware},
-         parent,
-         representation,
-         %{schema: schema, context: context} = resolution
-       ) do
+  defp resolve_reference_field(representation, resolution_acc) do
+    typename = Map.get(representation, "__typename")
+    %{schema: schema, source: source, context: context, adapter: adapter} = resolution_acc
     args = convert_keys_to_atom(representation, context)
 
-    middleware
-    |> maybe_unshim(schema)
-    |> Enum.find(nil, &only_resolver_middleware/1)
-    |> case do
-      {_, resolve_ref_func} when is_function(resolve_ref_func, 2) ->
-        fn _, _ -> resolve_ref_func.(args, resolution) end
-
-      {_, resolve_ref_func} when is_function(resolve_ref_func, 3) ->
-        fn _, _ -> resolve_ref_func.(parent, args, resolution) end
+    field =
+      schema
+      |> Absinthe.Schema.lookup_type(typename)
+      |> resolve_representation(representation)
+
+    %Absinthe.Resolution{
+      arguments: args,
+      schema: schema,
```

Comment:
I think this may be the last thing but I encountered some telemetry-related test errors in one of our codebases because these fields were missing on the resolution at the end. I didn't check what else can be taken from the resolution_acc and placed on the resolution we return here, but adding these two satisfied the test suite I was running. Also feel free to grab these in the pattern-match you added above if that's your preference.

```suggestion
      definition: resolution_acc.definition,
      parent_type: resolution_acc.parent_type,
      schema: schema,
```